Okayu enchantment

cpaca opened the pull request:
Added an enchantment for okayu. 

I'd go into more detail but I'd also like to sleep.

Finished implementation (only thing missing was for mods to change/approve recipe) on 2/22, so that's a :meow: in my books.

cpaca committed:
Implement BadAutoEat and AutoEat

See title.

Still a lot to do though. For now, they're exactly identical aside from the "when does it autoeat", though while testing I noticed they're both 10 levels to apply when autoeat probably shouldn't be 10 levels.

More importantly, though, the next commit will likely be about letting me make some Enchant's exclusive to other Enchant's. I'll need it anyway to make Haachama's enchant exclusive to Space Bread.

cpaca committed:
HoloEnchs *can* now be exclusive to each other.

That doesn't mean they have to, but they can be.

I looked everywhere where Enchant.exclusive is used (thanks, IntelliJ). I rewrote Utility.java's usage accordingly, and put comments in Events.java's 4 usages explaining why I didn't touch them.

I have not tested this code; the "test plan" for this is to try to put multiple AutoEat enchantments onto one helmet.

cpaca committed:
Generate basic AutoEat and GoodAutoEat code

Not implemented yet; I stopped because I realized maybe I should write that in Utility incase we want any other items to recursively search for an item in your inventory.

cpaca committed:
Implement Utility.recursiveSearchForItem

Yes, I had this code all done when I did the last commit. I just told GH desktop to not commit the last thing in the same commit, so you can look at the code for this one separately.

cpaca committed:
Implement AutoEat

Hmm... This Utility.recursiveSearchForItem function may need some reworking, though. I can change the ItemStack just fine but it doesn't update in the BlockMetaData.

cpaca committed:
Minor recursive change.

cpaca committed:
Implement GoodAutoEat, fix recursiveSearchForItem

I put an implNote that says what I think about this implementation.

cpaca committed:
Deactivate "all your items are 63 now"

cpaca committed:
Move food effects to Utility.java

cpaca committed:
Add lists of foods to Utility.java

I might need this later for recipes, or maybe it'll be deleted as better recipes are requested.

cpaca committed:
Add initial recipe ideas

This will probably end up changed later, but whatever.

cpaca committed:
Implement AutoEat exclusivity

That took a bit longer than I expected.

cpaca committed:
Remove debugging messages

cpaca committed:
Fix recipes.

REMINDER TO SELF:

If you ever, *ever* need to Fix Recipes again, make sure the HoloItem ingredients are in Events.ingredients.

cpaca committed:
File reorganization

cpaca committed:
Add Zero's recipe change suggestion

cpaca committed:
File renaming/deleting: Only one form of AutoEat

Only one form of AutoEat, which - since it's going on Okayu - I guess I'll name it MoguMogu.

This code doesn't work, but I at least fixed the recipes. Now to fix the rest of the stuff...

cpaca committed:
Override current code with upstream commit

Yes, I did just copy Enchant.java from upstream here. Hopefully that fixes the merge conflicts?

cpaca committed:
Merge branch 'main' of https://github.com/dlee13/OldHoloItems into dlee13-main

cpaca committed:
Merge pull request #2 from cpaca/dlee13-main

Successfully merge upstream commit.

cpaca committed:
Remove lingering instances of old AutoEats

cpaca committed:
Merge pull request #3 from cpaca/dlee13-main

Remove lingering instances of old AutoEats

cpaca committed:
Revert recursiveSearchForItem and addEnchant

cpaca committed:
Revert other weird Utility.java changes

cpaca committed:
No longer need food type separation

Since we're just using MoguMogu now.

cpaca committed:
Fix MoguMogu for HoloItems, sort of.

dlee13 commented on src/main/java/com/klin/holoItems/collections/gamers/okayuCollection/items/MoguMogu.java:
```diff
@@ -88,16 +89,40 @@ public static ItemStack autoEat(FoodLevelChangeEvent event) {
-        for(int i = 0; i <= 8; i++){
-            ItemStack item = inv.getItem(i);
-            if(item == null)
+        List<Integer> slots = new ArrayList<>(Arrays.asList(0,1,2,3,4,5,6,7,8,40));
```
`Arrays::asList` makes a list already, so doing `ArrayList::new` is redundant. Also, you can make it a `private static final` field called `SLOTS`.

dlee13 replied:
still unresolved

dlee13 commented on src/main/java/com/klin/holoItems/collections/gamers/okayuCollection/items/MoguMogu.java:
```diff
@@ -88,16 +89,40 @@ public static ItemStack autoEat(FoodLevelChangeEvent event) {
+            // Next, check if we can eat it and if we want to eat it.
+            Item holoItem = Utility.findItem(itemStack, Item.class);
+
+            if(holoItem == null) {
```
This predicate looks like it's the opposite of what is intended.

dlee13 commented on src/main/java/com/klin/holoItems/collections/gamers/okayuCollection/items/MoguMogu.java:
```diff
@@ -114,7 +139,7 @@ public static ItemStack autoEat(FoodLevelChangeEvent event) {
 
             // We're 100% eating this so
-            item.setAmount(item.getAmount() - 1);
+            itemStack.setAmount(itemStack.getAmount() - 1);
```
`ItemStack::subtract`

dlee13 commented on src/main/java/com/klin/holoItems/utility/Utility.java:
```diff
@@ -142,6 +142,75 @@ BlockFace.EAST, new Vector(1, 0, 0),
+     * Nutrition values from eating certain foods.
+     * Note: Chorus fruit and cake aren't listed here.
+     */
+    public static final Map<Material, float[]> nutritionValues = new HashMap<>(){{
```
`EnumMap`

dlee13 replied:
It's possible to get these values from the server internals by using reflection.
Relevant packages are:
net.minecraft.world.food
net.minecraft.world.item
We would want to access the foodProperties field of each Item.

The way to not hardcode this would be to iterate all materials, filtering for the ones that are edible and not legacy -> create a bukkit ItemStack for each -> convert to nms ItemStack -> `ItemStack::getItem` -> `Item::getFoodProperties` -> then the food properties instance has a bunch of getters for stuff.

cpaca replied:
"Convert to nms ItemStack" Ah, I was avoiding using NMS because I wasn't sure if we were allowed to have it, but since we are, this doesn't have to be hardcoded... might even be able to remove the variable entirely.

Reminder to self: Confirm what happens if you AutoEat Chorus Fruit.

dlee13 replied:
See discussion in discord

dlee13 replied:
We'd better move the Class and Method reflection stuff out to a new class so it'll be more maintainable and not such a mess. For iterating the Materials, it's gonna be:
```java
for (final Material material : Material.values()) {
    if (material.isEdible() && !material.isLegacy()) {
        ... // do the food property getting stuff and insert the values into the map
    }
}

dlee13 commented on src/main/java/com/klin/holoItems/utility/Utility.java:
```diff
@@ -142,6 +142,75 @@ BlockFace.EAST, new Vector(1, 0, 0),
+     * List of effects from eating certain foods.
+     * Note: Chorus fruit, rotten flesh, chicken, and honey bottles aren't listed here.
+     */
+    public static final Map<Material, PotionEffect[]> foodEatEffects = new HashMap<>(){{
```
`EnumMap`

dlee13 replied:
still unresolved

cpaca committed:
Complete almost all review changes

nutritionValues is still hard-coded though and I don't know how to fix that.

The instructions I was given were "convert to NMS ItemStack" and I don't know how.

cpaca committed:
Work to not-hardcode nutritionValues

Still not done, but it's progress.

Feels like black magic.

cpaca committed:
Make some variables final.

dlee13 commented on src/main/java/com/klin/holoItems/collections/gamers/okayuCollection/items/MoguMogu.java:
```diff
@@ -0,0 +1,187 @@
+
+    // Accepted armor types this can be on.
+    // Yes, both of these variables represent the same thing; it was written this way in DemonAura.
+    public static final Set<String> acceptedIds = Stream.of(ReadingGlasses.name, DragonHorns.name, Backdash.name).collect(Collectors.toCollection(HashSet::new));
```
Should be `Set::of`. This is allocating a stream for no reason. I know you want to say you did it this way because it's how Klin did it, but his coding style ignores modern java language features and makes the code harder to understand/maintain.

cpaca committed:
Fire PlayerItemConsumeEvent on HoloItems

cpaca committed:
Make nutritionValues not hard-coded

Note: This will eat chorus fruit but not teleport.

cpaca committed:
Forgot that saturation is capped at FoodLevel

cpaca committed:
Proper Frostfall Item Handling

cpaca committed:
[BROKEN] Uploading code so Termi can poke it

[BROKEN] This code is broken, but I'm uploading it so Termi can take a peek and see if I've missed something or if we'll come to the same conclusion

dlee13 commented on src/main/java/com/klin/holoItems/collections/gamers/okayuCollection/items/MoguMogu.java:
```diff
@@ -0,0 +1,211 @@
+    public MoguMogu(){
+        super(name, material, lore, durability, shiny, cost, acceptedIds, acceptedTypes, expCost);
+
+        // Incase I make an error like use BEETROOTS (block) insteaad of BEETROOT (item)
```
Could be made into a junit test instead

dlee13 replied:
On second thought, it'll be pointless after we make it not hardcoded.

dlee13 commented on src/main/java/com/klin/holoItems/utility/ReflectionUtils.java:
```diff
@@ -36,4 +86,20 @@ public static int getRepairCost(ItemStack stack) {
         return 0;
     }
+
+    /**
```
`ItemStack::finishUsingItem` handles everything, including decrementing the stack, incrementing stats, and applying effects

dlee13 commented on src/main/java/com/klin/holoItems/collections/gamers/okayuCollection/items/MoguMogu.java:
```diff
@@ -0,0 +1,211 @@
+                continue;
+            }
+
+            // We're eating this item.
```
After this point, it should only be the call to finishUsingItem. everything gets handled there by vanilla mechanics

cpaca committed:
Finalize all ReflectionUtils parts

cpaca committed:
Prepare the Black Magic

This is a separate commit to the next one incase I want to undo the next commit but keep the black magic that I've written here.

cpaca committed:
MoguMogu *almost* working with finishUsingItem

cpaca committed:
Now dodging the infinite loop

Unfortunately now the hunger and saturation aren't increasing, for some reason.

cpaca committed:
Make (some) requested changes

I didn't do the nutritionValues changes because I figured we wouldn't need it if I got finishUsingItem to work correctly.

On that note, finishUsingItem should have been fixed the past ... many commits
